src/persist-client/src/internal/compact.rs

```diff
@@ -248,6 +248,20 @@ where
         writer_id: WriterId,
     ) -> Result<CompactRes<T>, anyhow::Error> {
         let () = Self::validate_req(&req)?;
+
+        // special case: if our inputs only contain a single batch with >0 updates in a single run,
+        // we can return the same batch with an updated description to avoid pulling it down and
+        // rewriting it.
+        let filled_batches: Vec<_> = req.inputs.iter().filter(|x| x.len > 0).collect();
```

Comment:
I think we should figure out a way to avoid the Vec alloc when the special case doesn't apply. could remove the collect and manually call the iterator twice, something like

```
if let Some(batch) = filled_batches.next() {
  if let None = filled_batches.next() {
    return ...
  }
}
```

Reply:
we can do it in a single loop with an early-out, though I think I'm a little skeptical that one vec alloc per compaction is something to worry about

Reply:
iterators in rust are lazy, so my suggestion should amount to the same as you have here. 

>  though I think I'm a little skeptical that one vec alloc per compaction is something to worry about
fair. can't hurt now that we have it though